### PR TITLE
Add a more detailed modal when reporting a post or a comment

### PR DESCRIPTION
**Pull request by Flaburgan**
As a moderator,
I want to receive report only for posts and comments which violate the terms of use
To easily focus on the important ones

Many reports received on pods I moderate are a people whining about others "look how bad they talk to me". As a consequence of that flow, important reports are sometimes lost in the middle of this. As an attempt to reduce "not real report", I implemented a real modal (instead of the previous javascript alert) which tell to the users to report only content which violates the terms of use, and remind them that they can block the other user if they want to. Also, the reason is required but not filled, which force them to fill something.

Here is how it currently looks like:

report

### CODE REVIEW

File: app/assets/javascripts/app/views.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views.js]: Identifier 'item_id' is not in camel case.

File: app/assets/javascripts/app/views.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views.js]: Identifier 'item_type' is not in camel case.

File: app/assets/javascripts/app/views.js
Comment (Flaburgan) [inline app/assets/javascripts/app/views.js]: This is "old school javascript", I guess that function should be declared outside of the scope of this one, But I'm still not comfortable with backbone. Suggestions welcomed.

File: app/views/layouts/application.html.haml
Comment (Flaburgan) [inline app/views/layouts/application.html.haml]: The modal markup is needed on all page where there are posts, so streams and SPV. Is there a better place I can put it than in the global layout itself?

File: app/views/report/_report_modal.erb
Comment (Flaburgan) [inline app/views/report/_report_modal.erb]: I created an erb file instead of a haml one, thinking about a possible future where everything is translated in templates. ERB is HTML so the migration should be easier, and it's easier for me to write than HAML.

File: features/desktop/single_post_view_moderation.feature
Comment (Flaburgan) [inline features/desktop/single_post_view_moderation.feature]: I optimized that file by creating the post directly, not using the view. We won 11 seconds ;)

File: features/desktop/report.feature
Comment (Flaburgan) [inline features/desktop/report.feature]: I still don't succeed to go to that page. I have the error message:
`No route matches {:action=>"update", :controller=>"report"}, missing required keys: [:id] (ActionController::UrlGenerationError)`

I first thought it was because terms weren't enabled but I enable them in the Background section, and it doesn't change anything. I need help here.

**Comment by denschub** [comment]
> Many reports received on pods I moderate are a people whining about others "look how bad they talk to me". 

If your intention of this PR is to address this issue, then it won't help. Mainly because those who file those reports also would not read the modal. If you want to get people to report only valid things, then you have to go the Facebook/Twitter route and have them select from a list of predefined reason, some of which end in a dialog explaining how to block people, while others actually do submit a report.

**Comment by Flaburgan** [comment]
That's a good idea, but rise the question about how to make that generic for all pods... It means we could suggest a list the podmin would be able to edit, I guess.

**Comment by denschub** [comment]
I actually have the same question, but for your PR. Some pods, for example Geraspora, do not just refer to the ToS, but also to the project's community guidelines for content policing. Makes me wonder if we should expose that text in the config, but that would kill all chance of localizing that. Hm.

File: app/assets/javascripts/app/views.js
Comment (denschub) [inline app/assets/javascripts/app/views.js]: We should make sure to disallow sending reports without a reason submitted. I see a lot of reports using the default value in my database, and those reports are more or less useless. Forcing people to at least come up with an explanation is a good thing.

File: app/assets/stylesheets/interactions.scss
Comment (denschub) [inline app/assets/stylesheets/interactions.scss]: This looks like accidental white-space?

File: app/views/report/_report_modal.erb
Comment (denschub) [inline app/views/report/_report_modal.erb]: Can we make this a `textarea`? Single-line text fields always give the impression that the user is unable to write long explanations, and I'd really like them to be able to elaborate on their reports.

**Review by skyschub — CHANGES_REQUESTED** [review]
I have reconsidered my opinion here. I think referring to the terms is fine, as podmins can adjust their ToS templates to include links to additional documents like the Community Guidelines.

Some comments. I'm not really a fan of the current wording, and will propose an alternative once I had a bit more time to consider that. :)

**Comment by weex** [comment]
Bump on conflicting file

File: app/assets/javascripts/app/views.js
Comment (Flaburgan) [reply app/assets/javascripts/app/views.js:112]: The input has the `required` attribute, so any modern browser won't allow the user to submit the form if no reason is given. And it is empty by default. Do you think it is still worth doing a check in Javascript? Or even server side?

File: app/assets/stylesheets/interactions.scss
Comment (Flaburgan) [reply app/assets/stylesheets/interactions.scss:66]: Indeed, fixed.

File: features/desktop/report.feature
Comment (Flaburgan) [reply features/desktop/report.feature:25]: up on this, @SuperTux88 could you please explain to me:
- why there is no get route defined in routes.rb, but accessing /report currently works
- why capybara is trying to navigate to /report with the "update" action instead of doing a simple get?

File: app/assets/javascripts/app/views.js
Comment (denschub) [reply app/assets/javascripts/app/views.js:112]: > Or even server side?

*Always* validate state on the server side.

File: app/views/report/_report_modal.erb
Comment (Flaburgan) [reply app/views/report/_report_modal.erb]: Done.

**Comment by Flaburgan** [comment]
I rebased this PR, but I also marked it as Draft because of https://github.com/diaspora/diaspora/pull/8035#discussion_r296461071 if someone can explain this to me (@jhass maybe?)

File: app/assets/javascripts/app/views.js
Comment (Flaburgan) [reply app/assets/javascripts/app/views.js:112]: What I meant was, in that case, I didn't change anything on the server side. Anyway I just tried and it already fails if the report content is empty. So is server side handling + `required` in the markup enough, or do we want to double check on the client side, for very old browsers?

Comment (denschub) [reply app/assets/javascripts/app/views.js:112]: That sounds like a good-enough implementation, then. :)

**Comment by tclaus** [comment]
@Flaburgan Can you check how this will show in mobile view? The Issue "Reporting in mobile" is yelled several times and I also think a mobile 'reporting' is very useful.

**Review by tclaus — CHANGES_REQUESTED** [review]
Bildschirmfoto 2021-08-27 um 07 43 43


Some points I found: 

* The text-field is too high. 
* When reporting another Item, the old reporting text appears. textfield should be cleared.
* Is there a default / placeholder text? 
* Can you set the focus on the input-field after opening the dialog?

File: config/locales/diaspora/en.yml
Comment (tclaus) [reply config/locales/diaspora/en.yml:1043]: > If you want a place-holder, I suggest 'Enter your report here'. Or, you could enter "Please state how the content you are reporting violates our terms of use." from reason_label as a place-holder.

Exactly, a request to act, not just an empty space. 
Let us keep in mind (not for this PR) that other networks offer their reportes a list of (most used) topics. If nothing fit, user can give a text. 

For an analysis by moderators or admins, its good to keep this texts / categories short. But that's another Issue not to solve here.

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1048]: Ok done.

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1047]: Perfect.

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1046]: >My understanding is that ignoring doesn't prevent you from seeing comments on other people's posts; it just stops you from seeing that person's own posts, and stops them from interacting on your posts.

Yeah I just made the test and indeed the comment of an ignored user is still visible on someone else post. So I'm not sure about "you can hide it", as it is only true in some cases.

Comment (tclaus) [reply config/locales/diaspora/en.yml:1046]: Is this a bug or a feature? I would expect not to see anything if I ignore the user. Can you create an issue then?

Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1046]: I can't find of anything better right now, so I'll go with "If it doesn't but you don't want to see it, you can ignore the authors from their profile page. If the content is a comment on your own post, you can delete it directly" as the sentence is just after "Reporting content sends an email to all the moderators of this pod. Please only report content that violates pod_name terms" I think there is no need to repeat "violates ..."

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1043]: What about "This content violates the terms because..."? Placeholder are used to give an example of bad content, but I don't want to specify one as the terms are specific to the pod so I can't know what is forbidden, apart probably child porn but I don't want to mention this.

**Comment by Flaburgan** [comment]
Okay I think I fixed everything please tell me if this is now fine.

File: config/locales/diaspora/en.yml
Comment (goobertron) [reply config/locales/diaspora/en.yml]: You could replace the first sentence with "If you want to stop someone interacting on your posts, you can ignore them from their profile page."

It doesn't say it will hide all content, but gives an action you can do.

Comment (goobertron) [reply config/locales/diaspora/en.yml]: > Is this a bug or a feature?

@tclaus the ignore feature is not meant to be like Facebook's 'block' in which it's as if the other person no longer exists. It's called ignore because it's primarily a method of curating your stream; it means that their posts will no longer appear in your stream even if they contain tags you follow. It's primarily intended to deal with people who post a lot of stuff you're not interested in reading rather than harassment. That's why we should continue to refer to it as 'ignore' rather than 'block'.

**Comment by Flaburgan** [comment]
@SuperTux88 would you be able to tell me why the Jasmine tests are not running?

**Comment by Flaburgan** [comment]
@SuperTux88 if you know why jasmine is not running I think this is the only problem here

**Comment by SuperTux88** [comment]
@Flaburgan I have no idea why it's failing, it works perfectly locally. I rebased again on `develop` to remove the commits not belonging to this PR (so you want to "force pull" your local branch using `git reset` before you do any development again), and also added a commit to try to increase the timeout, however I have no idea if that helps. It would still be strange since it's usually finishing in under a minute, and if it suddenly takes more than two minutes, still something feels "broken" ... but lets see if it helps at all first, or if it's still timeouting. I think we added the timeout in the first place because when switching to GitHub actions we had problems with jasmine randomly getting stuck, but it's also strange that it now gets stuck with this PR consistently :man_shrugging:

**Comment by SuperTux88** [comment]
Ok, it didn't help, I'm out of ideas then ... I removed my commit again :man_shrugging:

**Comment by SuperTux88** [comment]
@Flaburgan If you change the `chrome_cli_options` you need to make sure that running it in docker still works (which needs the option currently on `develop`), so if GitHub-Actions need something special, then we need to set them conditionally (using an env-var for example). But I think the `chrome_cli_options` aren't the problem here, the options on `develop` work both in docker and on GitHub with the current `develop` code, so it's something in the code which breaks it, probably a change in `app/assets/javascripts/app/views.js`?

**Comment by Flaburgan** [comment]
When looking at the log in github, the only one we have is `[1001/011425.700187:ERROR:gpu_init.cc(453)] Passthrough is not supported, GL is disabled, ANGLE is`. When you search for it [on stackoverflow](https://stackoverflow.com/search?q=Passthrough+is+not+supported) everyone is saying that Chrome Driver crashes after that error, so I think something is misconfigured...

**Comment by SuperTux88** [comment]
The same error also is logged in `develop` where jasmine works fine, so I don't think that error is a problem (also if it would crash then it would just exit and not hang and timeout). I think it's more than one of the changes makes one of the tests run in an endless loop or something (no idea, just guessing), but it's strange that this only happens on GitHub. Maybe there is a way to give jasmine a more verbose output, to find where it's hanging?

**Comment by Flaburgan** [comment]
Ok I'm not available this week-end so this will have to wait for next week.

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml]: Oh yeah, we did the test one month ago in that exact thread :sweat_smile:

**Comment by Flaburgan** [comment]
@SuperTux88 looks like `--trace` isn't helpful on github. However, launching tests locally with `jasmine:ci` does hang on my computer. How could I know what's going on? Any way to display chrome logs in the console, or are they somewhere on my computer?

**Comment by Flaburgan** [comment]
Actually it fails even when accessing locahost:8888 saying /getting_started/ isn't found 404

**Comment by SuperTux88** [comment]
So yes, it looks like it's at least running a few tests, but then hangs. So probably not related to chrome errors at the beginning, since they are also printed when it's running and it also still runs a few tests now.

The last test which successfully runs is `app.Router _initializeStreamView sets app.publisher`, the next test that would run (when I run it locally it works) would be `app.Router _initializeStreamView doesn't set app.publisher if already defined`, so maybe that's the test which causes the hang? Maybe that helps you to debug that more.

**Comment by Flaburgan** [comment]
@SuperTux88 I tried things but nothing work, if you want I'm available this evening to check together

**Comment by Flaburgan** [comment]
Unfortunately the CI is still failing.

File: app/views/layouts/application.html.haml
Comment (diaspora-code-review) [inline app/views/layouts/application.html.haml:28]: Line contains trailing whitespace

**Review by SuperTux88 — CHANGES_REQUESTED** [review]

File: app/views/report/_report_modal.erb
Comment (SuperTux88) [reply app/views/report/_report_modal.erb:15]: See also https://github.com/diaspora/diaspora/pull/8239#discussion_r956804649 and https://github.com/diaspora/diaspora/pull/8239#discussion_r956804826

File: features/step_definitions/custom_web_steps.rb
Comment (Flaburgan) [reply features/step_definitions/custom_web_steps.rb:214]: Done.

File: app/assets/javascripts/app/app.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/app.js:162]: - Unexpected console statement.
- Missing semicolon.

File: app/assets/javascripts/app/app.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/app.js:164]: Multiple spaces found before 'ev'.

File: app/assets/javascripts/app/app.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/app.js:169]: Identifier 'item_id' is not in camel case.

File: app/assets/javascripts/app/app.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/app.js:170]: Identifier 'item_type' is not in camel case.

File: app/assets/javascripts/app/app.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/app.js:172]: Missing semicolon.

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1046]: So I tried to solve this with this line: `<%= t("report.modal.report_only_if_violated_terms", link_to_terms: link_to(t("report.modal.terms_link", pod_name: pod_name), terms_path, target: "_blank")) %>.` but then the HTML is not interpreted. Is this really what we want anyway?

Comment (SuperTux88) [reply config/locales/diaspora/en.yml:1046]: You need to mark the result as html safe with `.html_safe`, otherwise it gets escaped.

File: features/desktop/report.feature
Comment (Flaburgan) [inline features/desktop/report.feature:48]: I added a new test to ensure the correct post is reported, by the correct user.

File: app/assets/javascripts/app/app.js
Comment (Flaburgan) [reply app/assets/javascripts/app/app.js:168]: I'm not doing this refactor in this MR.

File: app/assets/javascripts/app/app.js
Comment (Flaburgan) [inline app/assets/javascripts/app/app.js:162]: So I changed the way we deal with the problem, the event is attached only once when the page is loaded, and the data about the post/comment is added to the form when the user click on the icon to open the modal. This ensures no multiple events, and that it's always the good post which is reported. Much cleaner.

File: app/views/report/_report_modal.erb
Comment (Flaburgan) [reply app/views/report/_report_modal.erb]: Changed this to add a sentence even when there are no terms.

File: app/views/report/_report_modal.erb
Comment (Flaburgan) [reply app/views/report/_report_modal.erb]: I don't think it really is important actually.

Comment (Flaburgan) [reply app/views/report/_report_modal.erb]: I think the whole interface can be improved, but because the work on this already started in https://github.com/diaspora/diaspora/pull/8239 I don't think I should change it in this MR, it will create conflicts. I think it's fine to merge it even if everything is displayed on one line, it doesn't break the UI: 
image
If you disagree then I would suggest to have only one merge request for this and https://github.com/diaspora/diaspora/pull/8239 so that we work on the same branch and are not overlapping the work.

**Comment by Flaburgan** [comment]
@SuperTux88 ok so code wise we should be good to go. Jasmine is still hanging sometimes, I think @denschub investigated this and his suggestion was to remove the problematic test? I don't remember and I don't know where his commit is.

**Comment by denschub** [comment]
> I think @denschub investigated this and his suggestion was to remove the problematic test? I don't remember and I don't know where his commit is.

Yes, see #8360. The broken test is

https://github.com/diaspora/diaspora/blob/03796e8fe2421ddfbc5380b8194ee68afbe3b88b/spec/javascripts/app/views/bookmarklet_view_spec.js#L46

and it's re-enabled because you force-pushed my commit away. :D You can re-disable the test, simply by replacing the `it(` with a `xit(`. As I said in the linked bug, I think it's fine disabling this test for now - it's not critical, and we have a tracking bug open to eventually fix it.

**Comment by Flaburgan** [comment]
>t's re-enabled because you force-pushed my commit away. :D 

Aaaaah sorry, indeed. I found back your commit https://github.com/diaspora/diaspora/commit/840478b2d2e5bd2162929b033426a18201c27b4b but even after fetching your fork I was not able to cherry-pick it. That's fine I just rewrote it. So, CI should be green soon, fingers crossed.

Do we want to disable the display of Jasmine output that we are introducing here or do we want to keep it?

**Comment by SuperTux88** [comment]
> Do we want to disable the display of Jasmine output that we are introducing here or do we want to keep it?

That commit was only for debugging and should be removed, same for the increase of the timeout (Both were already removed, but you brought them back when you removed the commit that disabled the test).

**Comment by Flaburgan** [comment]
Ok I will remove them. And https://github.com/diaspora/diaspora/pull/8035/commits/d63e3b94d825774a86a64935b704f7c23ef4eed5 ? Not sure how useful it is in the end...

**Comment by denschub** [comment]
If it works without, it's not useful. :) Our test suite doesn't do a .. god job at maintaining a consistent state for each test, so it might be possible that that's needed, but I'd try without first.

And if it's needed, please file a new bug to fix the tests that do weird things.

**Comment by SuperTux88** [comment]
> And https://github.com/diaspora/diaspora/commit/d63e3b94d825774a86a64935b704f7c23ef4eed5 ?

Oh yeah, that was already removed again too. So if it works without it, it can be removed too.

**Comment by Flaburgan** [comment]
ok, everything should be fine now. Only question is, do you agree with https://github.com/diaspora/diaspora/pull/8035#discussion_r1005526826 ?

File: app/views/report/_report_modal.erb
Comment (SuperTux88) [reply app/views/report/_report_modal.erb]: Well, it needs to be fixed somewhere, and I would have thought it fits better to this PR, because as soon as this is merged people are probably starting to send multi-line reports which then aren't rendered correctly. But if you rather want to fix it in the other PR, that's fine for me too, as long as both are in the same release (so 1.0 probably). But the current state is pretty "broken".

Comment (SuperTux88) [reply app/views/report/_report_modal.erb]: But if it's not working, it can also be removed again?

And it might not be super important, but it's still nice to have, and I think it shouldn't be hard to set focus with javascript when opening the modal?

File: app/views/layouts/application.html.haml
Comment (SuperTux88) [reply app/views/layouts/application.html.haml:44]: I don't know if you didn't see my previous comment here or if there is another reason, but you didn't change or comment anything here.

**Review by SuperTux88 — CHANGES_REQUESTED** [review]
Some minor things

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1048]: Inadequate? Unsuitable? Improper? cc @goobertron

Comment (goobertron) [reply config/locales/diaspora/en.yml:1048]: What is the purpose of `report_only_if_bad` that is different from `report_only_if_violated_terms`? What different type(s) of content is it intended to cover?

Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1048]: If the pod has terms enabled `report_only_if_violated_terms` is displayed otherwise we fallback on `report_only_if_bad`.

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml:1049]: Fair enough, removed ;)

File: app/views/report/_report_modal.erb
Comment (Flaburgan) [reply app/views/report/_report_modal.erb]: As said I would like to do this in https://github.com/diaspora/diaspora/pull/8239 to avoid conflicts. I will take care of it once this PR is merged if Thorsten is not around.

File: app/views/people/index.html.haml
Comment (diaspora-code-review) [inline app/views/people/index.html.haml:42]: Files should end with a trailing newline

File: app/views/people/show.html.haml
Comment (diaspora-code-review) [inline app/views/people/show.html.haml:48]: Files should end with a trailing newline

File: app/views/shared/_stream.haml
Comment (diaspora-code-review) [inline app/views/shared/_stream.haml:5]: Files should end with a trailing newline

File: config/locales/diaspora/en.yml
Comment (Flaburgan) [reply config/locales/diaspora/en.yml]: I replaced it with a link to the community guidelines.

File: app/assets/javascripts/app/views.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views.js:113]: Identifier 'item_id' is not in camel case.

File: app/assets/javascripts/app/views.js
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views.js:114]: Identifier 'item_type' is not in camel case.

File: app/views/people/index.html.haml
Comment (SuperTux88) [reply app/views/people/index.html.haml]: If you use an editor/IDE with editorconfig-support or with the respective plugin (https://editorconfig.org/), the editor would add the required newline automatically for you (https://github.com/diaspora/diaspora/blob/develop/.editorconfig#L5) :smiley:

**Comment by Flaburgan** [comment]
Everything is rebased and tests are green. @SuperTux88 if you can review it again, and once it's merged I can switch on https://github.com/diaspora/diaspora/pull/8239

**Review by SuperTux88 — APPROVED** [review]
Removed some extra newlines myself, otherwise looks good. Assuming you'll fix https://github.com/diaspora/diaspora/pull/8035#discussion_r956791512 in https://github.com/diaspora/diaspora/pull/8239

**Pull request merged**
